src/main/java/com/linkedin/kafka/clients/largemessage/LargeMessageSegment.java

```diff
@@ -16,54 +17,170 @@
+   *
+   * @param segment position should be assigned to the start of the segment, limit the end of the segment, remaining()
+   *                should give the number of bytes in the segment
+   */
   public LargeMessageSegment(UUID messageId,
                              int sequenceNumber,
                              int numberOfSegments,
-                             int messageSizeInBytes,
-                             ByteBuffer payload) {
-    this.messageId = messageId;
-    this.sequenceNumber = sequenceNumber;
-    this.numberOfSegments = numberOfSegments;
-    this.messageSizeInBytes = messageSizeInBytes;
-    this.payload = payload;
+                             int originalValueSize,
+                             boolean originalKeyWasNull,
+                             ByteBuffer segment) {
+    if (messageId == null) {
+      throw new InvalidSegmentException("messageId must not be null.");
+    }
+    this._messageId = messageId;
+
+    if (sequenceNumber < 0) {
+      throw new InvalidSegmentException("sequenceNumber must be non-negative.");
+    }
+    this._sequenceNumber = sequenceNumber;
+
+    if (sequenceNumber >= numberOfSegments || numberOfSegments < 0) {
+      throw new InvalidSegmentException("Sequence number " + sequenceNumber
+        + " should fall between [0," + (numberOfSegments - 1) + "].");
+    }
+    this._numberOfSegments = numberOfSegments;
+
+    if (originalValueSize < 0) {
+      throw new InvalidSegmentException("originalValueSize must be non-negative.");
+    }
+    this._originalValueSize = originalValueSize;
+
+    if (segment == null) {
+      throw new InvalidSegmentException("segment ByteBuffer must not be null");
+    }
+    this._segment = segment;
+
+    this._originalKeyWasNull = originalKeyWasNull;
   }
 
   /**
-   * Notice that the payload as a ByteBuffer does not guarantee to have a dedicated underlying byte array. So calling
-   * {@code payload.array()} will not always give the payload byte array. This method should be called if user wants
-   * to have the payload byte array.
+   * This constructor should be used when consuming a record from the broker.
    *
-   * @return The payload as a byte array.
+   * @param headerPayload non-null, this was originally generated from the producer side.  This is stored in the header
+   *                      map with key {@value HeaderKeySpace#LARGE_MESSAGE_SEGMENT_HEADER}
+   * @param segment the bytes for this segment
    */
-  public byte[] payloadArray() {
-    if (payload.arrayOffset() == 0 && payload.limit() == payload.array().length) {
-      return payload.array();
-    } else {
-      return Arrays.copyOfRange(payload.array(), payload.arrayOffset(), payload.arrayOffset() + payload.limit());
+  public LargeMessageSegment(byte[] headerPayload, ByteBuffer segment) {
+    ByteBuffer headerReader = ByteBuffer.wrap(headerPayload);
+    byte headerVersion = headerReader.get();
+    if (headerVersion != CURRENT_VERSION) {
```

Comment:
In open source there will not be other versions.  Internally we can have our own implementation that recognizes the old version and either produces an empty map or a single header which describes the large message segment.